pkg/execution/batch/redis.go

```diff
@@ -133,14 +136,28 @@ func (b redisBatchManager) Append(ctx context.Context, bi BatchItem, fn inngest.
 	return result, nil
 }
 
+func (b redisBatchManager) isSharded(ctx context.Context, batchPointer string, functionId uuid.UUID) (redis_state.RetriableClient, bool) {
+	// Check whether passed batch pointer key includes sharded queue prefix
+	isSharded := strings.HasPrefix(batchPointer, b.b.KeyGenerator().QueuePrefix(ctx, functionId))
+
+	client := b.b.UnshardedClient()
+	if isSharded {
+		client = b.b.ShardedClient()
+	}
+
+	return client, isSharded
+}
+
 // RetrieveItems retrieve the data associated with the specified batch.
-func (b redisBatchManager) RetrieveItems(ctx context.Context, batchID ulid.ULID) ([]BatchItem, error) {
+func (b redisBatchManager) RetrieveItems(ctx context.Context, functionId uuid.UUID, batchID ulid.ULID, batchPointer string) ([]BatchItem, error) {
```

Comment:
Retrieving can happen immediately after `Append` when the batch is full, or after the batch timeout has passed and execution is started. We need to check whether we're dealing with an old (unsharded) batch or a new (sharded) batch. To find out, we can simply compare the prefix with the sharded queue prefix.